Rover/JetsonGStreamerServer/gserver.cpp

```diff
@@ -0,0 +1,227 @@
+  scale_cap = gst_element_factory_make("capsfilter", NULL);
+
+  // options for the source
+  g_object_set(G_OBJECT(source), "device", gst_structure_get_string(str, "device.path")
+                                                                                , NULL);
+  std::unordered_map<std::string, std::tuple<gint, gint, gint, gint, gint> > table;
+  table["RICOH THETA S"] = std::make_tuple(1280, 720, 5555, 640, 360);
+  table["ZED"] = std::make_tuple(3840, 1080, 5556, 960, 270);
+  table["USB 2.0 Camera"] = std::make_tuple(1280, 720, 5557, 640, 360);
+ 
+  
+  s_cap_unapplied = gst_caps_new_simple("video/x-raw", "format", G_TYPE_STRING, "I420",
+                              "width", G_TYPE_INT, std::get<0>(table[name]), "height", 
+                               G_TYPE_INT, std::get<1>(table[name]), NULL);
+  g_object_set(G_OBJECT(s_cap), "caps", s_cap_unapplied, NULL);
+
+  // options for the encoder
+  e_cap_unapplied = gst_caps_new_simple("video/x-h264", "stream-format", 
+                                     G_TYPE_STRING, "byte-stream", NULL);
+  g_object_set(G_OBJECT(e_cap), "caps", e_cap_unapplied, NULL);
+
+  // options for the videoscalar
+  scale_cap_unapplied = gst_caps_new_simple("video/x-raw", "format", G_TYPE_STRING,"I420", 
+                                    "width", G_TYPE_INT, std::get<3>(table[name]), 
+                                    "height", G_TYPE_INT, std::get<4>(table[name]), NULL);
+ 
+  g_object_set(G_OBJECT(scale_cap), "caps", scale_cap_unapplied, NULL);
+
+  // options for the sink
+  g_object_set(G_OBJECT(sink), "host", client, "port", std::get<2>(table[name]), NULL);
+  if (!pipeline || !source || !s_cap || !e_cap || !depay || !encoder || !parse || !sink || !scale || !scale_cap) 
+  {
+    g_printerr ("One element could not be created. SAD! \n");
+    exit(-1);
+  } 
+  
+  // add elements and linking them
+  gst_bin_add_many(GST_BIN(pipeline), source, s_cap, e_cap, depay, 
+                    encoder, parse, sink, scale, scale_cap, NULL);
+ 
+  if (!gst_element_link_many(source, s_cap, scale, scale_cap, encoder, 
+                                     e_cap, parse, depay, sink, NULL)) 
+  {
+    g_printerr ("unable to link the elments to the pipeline. SAD! \n");
+    exit(-1);
+  } 
+
+  return pipeline;
+}
+
+GstDeviceMonitor * device_monitor (void) 
+{
+  GstDeviceMonitor *monitor;
+  GstBus *bus;
+  GstCaps *caps;
+
+  // starts the monitor for the devices 
+  monitor = gst_device_monitor_new ();
+
+  // starts the bus for the monitor
+  bus = gst_device_monitor_get_bus (monitor);
+  gst_object_unref (bus);
+
+  // adds a filter for the devices seen
+  caps = gst_caps_new_empty_simple ("video/x-raw");
+  gst_device_monitor_add_filter (monitor, "Video/Source", caps);
+  gst_caps_unref (caps);
+
+  gst_device_monitor_start (monitor);
+
+  return monitor;
+}
+
+// creates the pipelines and uses them in different threads
+void stream_start(GList *cur, int argc, char *argv[]) {
```

Comment:
STYLE POINTS, by which I mean be consistent with your curly braces. Pick one or the other (I prefer next line).